### PR TITLE
Split `function` into `constructor`/`relation`/(custom)`function`; Remove `default`; Disallow `function` lookup in the RHS of a rule

### PR DESCRIPTION
**Pull request by FTRobbin**
This PR fixes Issue #420. Lookup actions in rules will now cause a type error `LookupInRuleDisallowed`.

Move specifically, this PR:

1. Removes `-naive` flag and related desugaring code due to being replaced by this change.

2. Fixes 'fail' failing due to not being identified as global in the `remove_global` rewrite pass.

3. Adds new positive and negative tests for this type error.

4. Rewrites the existing tests for compatibility with the new type error.

### CODE REVIEW

File: src/gj.rs
Comment (yihozhang) [inline src/gj.rs:755]: I believe we still want to keep the `-naive` flag as well as the code here, so the user can still do naive evaluation (useful for debugging, also have a different semantics than semi-naive for "unsafe" egglog).

**Comment by saulshanabrook** [comment]
I'm a little worried about the time improvements, especially for lambda... That one is so dramatic I worry that maybe the semantics of the example changed?

Seeing all the changes, I also worry about the degradation for UX, it seems just more unwieldy with this change. 

I know you said that automated desuguring had some issues, but I am wondering if that could be used to at least addressost of these cases? Where there particular issues with it for some cases or just in general?

File: src/typechecking.rs
Comment (oflatt) [inline src/typechecking.rs:373]: Don't you need to check if this is a function vs a constructor call here?

File: tests/eggcc-extraction.egg
Comment (oflatt) [inline tests/eggcc-extraction.egg:1529]: `IntI` is a constructor, not a function
So this isn't a lookup and doesn't need to be changed

**Comment by FTRobbin** [comment]
Bumping the PR for review. Now, it properly deals with Issue #420 by enforcing a write-only RHS for rules through a new type-checking pass that checks every subexpression.

This restriction forbids looking up functions even with a default value. This breaks existing tests primarily due to the pervasive use of `relation`. I have manually rewritten the tests to use `set` instead.

I also added a few more test cases for this new pass.

I'm working on #421 to remove default values and #422 to distinguish constructors and functions. Due to the dependencies, I recommend merging after those two are done in this branch.

Known issues:
- merge functions can cause implicit read
- primitives can have side effects (uncertain)

File: tests/array.egg
Comment (oflatt) [inline tests/array.egg:23]: Oh I see, now relations have to be set. My opinion is that we shouldn't make users use `set` for relations.

I think the long-term plan is for us to desugar relations to be constructors so people don't have to do this. You could either implement that plan in this PR, or desugar relations to a `set` for now and do that in another PR.

**Review by oflatt — CHANGES_REQUESTED** [review]
Great job on this PR! I had some concerns that need to be addressed before we merge this- the main one has to do with `relation` desugaring

File: src/gj.rs
Comment (FTRobbin) [reply src/gj.rs:755]: I don't think we should keep it. It is unhelpful and adds complexity to the later passes as they need to support the naive semantics correctly. I am also against keeping it as a use-at-your-own-risk feature.

For Egglog users, if you don't use delete, semi-naive and naive are indistinguishable, so it is unhelpful for debugging. If you use delete, then you care much about performance, and there's no point in using naive. Even when you debug with unsafe features, you should probably debug the semi-naive case instead because that's what you want.

For Egglog developers, I see some value in being a sanity check for ensuring semi-naive is implemented correctly. But we are not doing this now, and it can also be done through stronger end-to-end test cases.

File: tests/array.egg
Comment (FTRobbin) [reply tests/array.egg:23]: I don't think I'll implement the desugaring in this PR. My impression is that it is still up to discussion what it should be desugared into. Performance overhead aside, desugaring into a constructor might not be ideal because that makes it more expressive than needed. For instance, now you can union two edges in the relation or have a relation as the codomain of a function.

Comment (oflatt) [reply tests/array.egg:23]: That's a good point- we would have to disallow unioning two relation terms.
You can't have a relation as the codomain of a function because the sort would be fresh (no way to get a handle on it)

Relation desugaring aside, I don't think we want to make this breaking change when we plan to fix it right away.

File: src/gj.rs
Comment (oflatt) [reply src/gj.rs:755]: That was convincing to me. I've never personally used it before
@yihozhang what do you think?

Comment (yihozhang) [reply src/gj.rs:755]: What complexity does this add to later passes? I thought the only difference between seminaive and naive is that in the seminaive case, we split the original query into many small queries depending on the timestamps (i.e., what this code snippet does).

I strongly recommend that we keep the naive evaluation. We can view semi-naive as an optimization of the naive evaluation, and this optimization is not always semantic-preserving, when given bizarre programs that violate certain assumptions. Examples include

* rules that use `extract` / user-defined primitives
* rules where the merge function is not associative or idempotent

I'm also not confident that our semi-naive is implemented correctly- do we really update timestamp every time we update the table? I just looked at table.rs and it seems we don't update the timestamp for at least `get_mut`. The naive evaluation serves as a ground truth for this purpose. Personally, when I am debugging a primitive I wrote, the first thing I do is to disable semi-naive evaluation.

File: tests/array.egg
Comment (yihozhang) [reply tests/array.egg:23]: Let's make functions whose output is `Unit` a special case: Such functions have default value `()` and an implicit merge function.

**Review by yihozhang — CHANGES_REQUESTED** [review]
Great PR!

File: tests/array.egg
Comment (FTRobbin) [reply tests/array.egg:23]: I don't think that would work. I think the plan is to get rid of default values altogether. Desugared or not.

Comment (yihozhang) [reply tests/array.egg:23]: I meant getting rid of the `default` field/keyword, but handle unit specially (similar to how EqSort is handled specially).

File: src/gj.rs
Comment (FTRobbin) [reply src/gj.rs:755]: If we keep the naive flag, either we need to split the latter passes into two, which is unlikely, or each piece of downstream code must support both naive and semi-naive. I am skeptical about the claim that semi-naive code would just work for naive. For one thing, I don't see how semi-naive can be implemented as pure syntactic rewrites. As you pointed out, something more needs to happen to the timestamps in the semi-naive case. And yet, the naive flag is not used anywhere else in the codebase.

There are cases where the two give different semantics. Still, the naive semantics is not more helpful to the users in those cases.

For your last point: Firstly, you still need to debug your new primitive work with semi-naive. Secondly, I will only trust naive evaluation as a ground truth if it is well supported with a clear separation between the two semantics. Relying on your program to be tested to produce test output is a terrible idea to me.

However, I do think this discussion raised a significant concern about the correctness of Egglog. We should investigate the issue.

File: tests/array.egg
Comment (FTRobbin) [reply tests/array.egg:23]: Conclusions:
- Conceptually: 3 function subtypes
  - Constructor | Relation | Custom
  - Operates differently
- Syntax
  - Keeping the status quo
    - Because there is not enough reason to change it
    - The decision is final for the June vision
  - Adding constructor keyword
    - Only relation if declared as relation
    - Only constructor if declared as constructor
- Semantics/implementation
  - Everything still uses the same function data structure under the hood
  - Only relation and constructor are allowed on RHS
     - It adds the relation/creates the term
  - Union requires both arguments to be the same EqSort
  - Set requires the head to be a custom function

File: src/gj.rs
Comment (FTRobbin) [reply src/gj.rs:755]: Conclusion: Keep
- Not comprising the comfort of `-naive` for a smaller core
- Too much effort to actually implement `-naive`, we settle for the timestamp hack
- Reconsider when merging the new backend

**Comment by FTRobbin** [comment]
Bumping up this PR again for review:

- Removed `default` keyword, resolving #421.
- Reverted the `-naive` flag change as discussed.
- Implemented splitting `function` into three subtypes: `constructor`/`relation`/(custom)`function`, resolving #420 & #422.
  - `function` is not allowed in the RHS of a rule (`merge` functions are unchecked).
    - `function` can have eqSort as an output.
      - It does not have union as the default merge function
  - `constructor` and `relation` are allowed
    - A `constructor` expression inserts a new enode
      - It has union as the default merge function
    - A `relation` expression inserts a new edge
  - A `constructor`’s output type must be sort
- Reverted the previous changes to tests and then fixed all the tests again
- Added new negative and positive tests

File: src/actions.rs
Comment (yihozhang) [inline src/actions.rs:294]: Nit: this should probably provide a different error message given this PR, since the only case this is possible is when there is a bug in our checker.

File: src/ast/mod.rs
Comment (yihozhang) [inline src/ast/mod.rs:366]: nice documentation!

File: src/ast/mod.rs
Comment (yihozhang) [inline src/ast/mod.rs:375]: The map part of the definition is a bit weird to me, but it's fine for now since we will need a big documentation refactor anyway.

File: src/ast/mod.rs
Comment (yihozhang) [inline src/ast/mod.rs:417]: nit Constructor -> constructor

File: src/ast/mod.rs
Comment (yihozhang) [inline src/ast/mod.rs:463]: nit: this is not an example where a custom function has an EqSort output

**Review by yihozhang — APPROVED** [review]
Nice job! Code is clean, with detailed documentation and good tests.

Let us make a release after this PR is merged.

File: tests/cykjson.egg
Comment (yihozhang) [inline tests/cykjson.egg:16]: Should B be a constructor so that `union` would still work?

File: tests/intersection.egg
Comment (yihozhang) [inline tests/intersection.egg:30]: this could just be `union`?

File: tests/lambda.egg
Comment (yihozhang) [inline tests/lambda.egg:65]: Same here. It's an interesting choice to make evals-to a custom function instead of a constructor. This is a new pattern to me, but it seems to work.

File: src/actions.rs
Comment (FTRobbin) [reply src/actions.rs:294]: I think it can still be triggered by merge functions reading a table, e.g.:

```
(function foo () i64)

(function bar () i64 :merge (foo))

(set (bar) 0)

(fail (set (bar) 1))
```

File: tests/lambda.egg
Comment (FTRobbin) [reply tests/lambda.egg:65]: This is indeed a new pattern. I'll explain more during the meeting.

File: src/actions.rs
Comment (oflatt) [reply src/actions.rs:294]: Yeah- could add a TODO since this is a known issue
In the future it should be checked and disallowed

File: tests/intersection.egg
Comment (oflatt) [reply tests/intersection.egg]: Are we disallowing `set` instead of `union` in a follow up PR?

File: tests/unification-points-to.egg
Comment (oflatt) [inline tests/unification-points-to.egg:147]: Should these be `union`?

File: src/ast/parse.rs
Comment (oflatt) [inline src/ast/parse.rs:372]: This doesn't seem right- shouldn't `:merge` be required for functions?
Otherwise, you can use a constructor (or panic for primitives).
Perhaps this should be a follow-up PR.

File: src/ast/mod.rs
Comment (oflatt) [inline src/ast/mod.rs:103]: is this "change" a todo?

File: src/ast/mod.rs
Comment (oflatt) [inline src/ast/mod.rs:117]: Why call this "Custom"? Why not name it "Function"? Are there other non-function things this could be?

File: src/ast/mod.rs
Comment (oflatt) [inline src/ast/mod.rs:373]: Nice- could add "relations are like constructors, but can't be `union`-ed

File: src/ast/mod.rs
Comment (oflatt) [inline src/ast/mod.rs:1148]: why does this say not used? Shouldn't clippy catch that problem?

File: src/typechecking.rs
Comment (oflatt) [inline src/typechecking.rs:400]: In the future, let's do this check on merge expressions (:

**Review by oflatt — CHANGES_REQUESTED** [review]
Awesome work Haobin! None of my comments are blocking, but they probably require follow-up PR(s)
Will approve once you respond to them

File: src/ast/parse.rs
Comment (FTRobbin) [reply src/ast/parse.rs:372]: This is not enforced right now. Functions without a merge function will throw an error when set with different values.

File: src/ast/mod.rs
Comment (FTRobbin) [reply src/ast/mod.rs:103]: The case here is that this change broke the one-to-one correspondence between `GenericCommand` and `GenericNCommand`. I'm not sure what to do with that.

File: src/ast/mod.rs
Comment (FTRobbin) [reply src/ast/mod.rs:117]: We decided to call it "custom function" to distinguish it from "function," which contains all three subtypes.

File: src/ast/mod.rs
Comment (FTRobbin) [reply src/ast/mod.rs:1148]: I think it is never used now. The old code used this as a catch-all for non-relation functions. Now, the three subtypes are handled separately. However, it is still there to implement `Display`.

Comment (oflatt) [reply src/ast/mod.rs:1148]: If it's there to implement display then it's used
Perhaps change the comment to "used in implementation of `Display`"

File: src/ast/mod.rs
Comment (oflatt) [reply src/ast/mod.rs:117]: ah ok

File: src/ast/mod.rs
Comment (oflatt) [reply src/ast/mod.rs:103]: I think it's okay, just wondering what you meant here

File: src/ast/parse.rs
Comment (oflatt) [reply src/ast/parse.rs:372]: Ah okay- so if we don't default to `union` even for eq-able types? That was the old behavior that was confusing

File: tests/unification-points-to.egg
Comment (oflatt) [reply tests/unification-points-to.egg]: Shouldn't we disallow `set` on constructors?

**Review by oflatt — APPROVED** [review]
🎉

**Comment by saulshanabrook** [comment]
I am finally implementing this in Python land and have noticed some weird runtime behavior where certain things are allowed that probably shouldn't be? Like you can union a function that outputs a primtive with a primtive and it can silently pass but just not do anything: https://github.com/egraphs-good/egglog/issues/483#issuecomment-2661487974

I was trying to make the Python types more narrow so that statically you would know like you can't union two primitives. But then I was trying to understand what the rules were and found that there doesn't seem to be many currently...

**Pull request merged**
